### ISSUE TITLE
A way to stop shrinking from within property evaluation ?

### ISSUE THREAD

**Issue by altxt** [hedgehogqa/fsharp-hedgehog#316]
When writing my first property tests with Hedghog, I bumped into SUT entering an endless loop. I solved this with running the offending function in a new Task and cancelling it on timeout via Thread.Abort and/or CancellationToken. In FsCheck this works fine for reasonable timeout durations. Hedgehog, however, treats timeout as a counterexample and starts trying to shrink it. Often this leads to another timeout. For 5 seconds of timeout and shrink limit of 20 it takes more that a minute. Depending on the property and the system, it might take a lot longer.

A developer debugging a property can handle this by adjusting the timeout value and the shrink limit. A CI server, on the other hand, will waste potentially a lot of time in attempts to shrink a hanging property, if the hanging happened in place where no one expected it. Disabling shrinking or setting the limit low enough seems to be a bad idea, as automatic shrinking is the reason I'm trying Hedgehog in the first place.

Right now, I solved this by detecting if the test is running on a CI server and enabling shrinking based on that. But it would be better if there was a way of telling Hedgehog that in some cases, detected at run time, further shrinking is to be skipped. One way to do this is by using a custom StopShrinkingException, but there may be other ways.

**Comment by TysonMN** [comment]
Do you have desired behavior in mind or are you only describing a use case for which we might be able to improve the behavior?

> A developer debugging a property can handle this by adjusting the timeout value and the shrink limit. A CI server, on the other hand, will waste potentially a lot of time in attempts to shrink a hanging property, if the hanging happened in place where no one expected it.

I understand general desire to minimize usage of CI server resources, but the primary goal is to save development time.  People are expensive; computers are cheap.

Can you rephrase this in terms of how you want to save development time instead of CI server time?

**Comment by altxt** [comment]
> I understand general desire to minimize usage of CI server resources, but the primary goal is to save development time. People are expensive; computers are cheap.

I'm not all that worried about server usage. Human time is actually more affected. PBT might be used not only as an R&D tool, but also as a guard against regressions, like traditional example- and script-based tests. So a CI build and test run that unexpectedly took too long might block a PR from merging, forcing developers to wait idly or multitask, neither of which is good for productivity.

Realistically, I think a CI test has to run on a time budget, so I'll probably just disable shrinking as I said above. But there might be value in having the problem stated in the logs clearly, which means clean data, which means shrinking.

> Do you have desired behavior in mind or are you only describing a use case for which we might be able to improve the behavior?
How about this? When a property evaluation encounters a kind of problem that justifies human intervention, not automatic simplification, it sends a signal to Hedghog, and the property fails right away.

**Comment by TysonMN** [comment]
What would a human do differently than what Hedgehog is already doing automatically?

**Comment by altxt** [comment]
After thinking some more, I believe the argument about time limit applies equally to a CI environment and to a local test run. Most of the time people cannot afford to wait an indefinite amount of time. Either a test completes in the available time span, or it has to be aborted (thus producing no usable output) and scheduled to run overnight, on weekend, etc.

The difference is in whether a developer is available to supervise the test run, look into the problems, and generally be an active participant.
If he is, the developer can choose a more whitebox approach and look into the code tested in the property, or check if the property itself is faulty, unstable or just not true. This includes playing with test count and shrink limit.
Or he can take a higher-level view, write the problem down, shelve it for later and move on, disabling the test.

If, on the opposite, the tests are supposed to help and safeguard the development effort, and just to "always pass" like typical unit tests are, then indeterminate run time starts to become a problem.

**Comment by TysonMN** [comment]
> Hedgehog, however, treats timeout as a counterexample and starts trying to shrink it. Often this leads to another timeout. For 5 seconds of timeout and shrink limit of 20 it takes more that a minute. Depending on the property and the system, it might take a lot longer.

I think there are at least two issues here.

First, it seems that fives seconds is too long.  I get the impression that you would prefer a lower timeout.  I don't have first-hand experience with Hedgehog's timeout behavior and API.  I think it would be great to be able to set a default timeout as well as local overrides.  Not sure if that is possible.

Is there some improvement to Hedgehog's "timeout API" that would make things better for you?

Second, I think you are saying that it might be possible to reach a shrunken input (i.e. an input such that all "smaller" inputs cause the test to pass) in fewer shrinks.  That is indeed a legitimate concern.

I significantly reduced the number of shrinks used in PR #239.  A new release has not been created since that PR was completed.  You can tell @moodmosaic or @adam-becker that you want a new release so that you can use that improvement.

Beyond that, there is a last-standing issue (#223) and a new one (#319, which I just created) that are about specific ways that a shrunken input can be found with fewer shrinks.  I plan to work on those two issues (eventually).

**Comment by altxt** [comment]
> First, it seems that fives seconds is too long.
I was referring to the running time of the function called within the property, not the Hedgehog machinery on top of that. 5 seconds was the threshold between a successful execution and a hung one for a particular test I was working with at the time. I'm not sure how long shrinking takes in addition to property evaluation, but it has to be time-limited all the same.

> Is there some improvement to Hedgehog's "timeout API" that would make things better for you?
Looking at how minimalistic Hedgehog is (there is no Collect or Classify, for example), and that there is more than one way to define a timeout (per SUT function call, per property, per test suite), I begin to doubt that it needs to be part of Hedgehog. Perhaps the timeout logic can be implemented separately (by test authors, that is) and simply invoke Hedgehog "interrupt API". The interrupt API needs only one signal - "stop doing stuff with the current property as soon as the library code regains control". It can be implemented with a custom exception type or by passing a CancellationToken to Property.Check.

**Comment by altxt** [comment]
As an afterthought, an interrupted property check should dump the last used seed and size so that it's possible to restart from where the test stopped.

**Comment by TysonMN** [comment]
> Looking at how minimalistic Hedgehog is (there is no Collect or Classify, for example) [...]

What would `Collect` and `Classify` do?

**Comment by altxt** [comment]
They are used to look at the generated input values, gather some statistics and see if they are good enough to really check what the property claims it does.
https://fscheck.github.io/FsCheck/Properties.html#Observing-Test-Case-Distribution
Collect just prints the value, so it can be replaced with a simple Console.WriteLine within a property.
Classify takes a bit more code to implement, but it can just as well be done independently of the test framework.

**Comment by TysonMN** [comment]
Ah, yes.  Thanks for that clarification.  Issue #25 is about adding such features.

> 5 seconds was the threshold between a successful execution and a hung one for a particular test I was working with at the time.

When the test passes, how long is a typically execution?

**Comment by altxt** [comment]
500ms - 1sec

**Comment by TysonMN** [comment]
I thought of a potential workaround.  Can you try including a filter in your generator that depends on a flag that you toggle after concluding that you want shrinking to stop?  Mutations are nasty though, so you might run into weird behavior like that in #192.
